python/ray/data/datasource/file_based_datasource.py

```diff
@@ -190,6 +198,16 @@ def __init__(
                     "'partition_filter' field is set properly."
                 )
 
+        if file_extensions is not None:
+            path_to_size = dict(zip(paths, file_sizes))
+            paths = [p for p in paths if _has_file_extension(p, file_extensions)]
+            file_sizes = [path_to_size[p] for p in paths]
+            if len(paths) == 0:
+                raise ValueError(
+                    "No input files found to read. Please double check that "
+                    "'file_extensions' field is set properly."
+                )
```

Comment:
This is an O(n) operation. If we want to optimize this in the future, we can pushdown the filtering to the metadata provider.

Reply:
maybe we can leave a todo comment so we don't lose it in this PR?